### ISSUE TITLE
regression: forced packing of non-POD leads to compatibility problems

### ISSUE THREAD

**Issue by stsp** [llvm/llvm-project#62358]
```cpp
#include <cstdio>
#define PACKED __attribute__((packed))

class Foo {
    int i;
public:
    int& get_ref() { return i; }
};

class Bar {
public:
    unsigned char c;
    Foo p;
    unsigned char c1;
    int c2;
    int get_ref() { return p.get_ref(); }
} PACKED;

int main()
{
  Bar b;
  printf("sizeof(b)=%zi\n", sizeof(b));
  return b.get_ref();
}
```

This code builds well on Clang 15, and prints the result "10", which is correct.
But on Clang 16 we get this:
```console
$ clang++-16 -Wall -Wpacked-non-pod -Wpacked -o p_clng16 pod.cpp
pod.cpp:13:9: warning: not packing field 'p' as it is non-POD for the purposes of layout [-Wpacked-non-pod]
    Foo p;
        ^
1 warning generated.
```
and the result is 16, which is not what we need. So the work-around is to pack class `Foo`, lets do that. Now we get:
```console
$ clang++-16 -Wall -Wpacked-non-pod -Wpacked -o p_clng16 pod.cpp
$ ./p_clng16
sizeof(b)=10
```

So far, the work-around works as expected, good. But suddenly we discover this:
```console
$ g++ -Wall -Wpacked -o p_gcc pod.cpp
pod.cpp: In member function ‘int& Foo::get_ref()’:
pod.cpp:7:29: error: cannot bind packed field ‘((Foo*)this)->Foo::i’ to ‘int&’
    7 |     int& get_ref() { return i; }
```

So with such work-around, the code no longer builds with GCC. See here for details: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=36566

So by forcing people to pack whatever is already packed, we get the deserved punishment from `-Wpacked` (#62353 ) and also get the code that can't be compiled with GCC, and works properly with Clang only because Clang allows to bind packed fields to a reference. I wouldn't be surprised if one day Clang would disallow also that, and the
work-around of adding explicit packing will stop working even for Clang.

The problem is introduced by https://github.com/llvm/llvm-project/commit/277123376ce08c98b07c154bf83e4092a5d4d3c6

**Comment by shafik** [comment]
CC @dwblaikie

**Comment by dwblaikie** [comment]
Working through this... 

> This code builds well on Clang 15, and prints the result "10", which is correct.

what is "correct" is open to debate - it's what clang previously determine as the size, but not what GCC produces: https://godbolt.org/z/5ee134aKs

> So with such work-around, the code no longer builds with GCC. 

But with GCC you weren't getting the packing you wanted previously anyway (you were getting `sizeof(Bar) == 16`, right? - so, if under GCC you were OK without packing the non-pod member, then you could conditionalize the struct packing - or take the error from GCC as a sign it's diagnosed a bug that existed in your code but Clang didn't diagnose - and you should use the size 16 struct layout instead of size 10)

> So by forcing people to pack whatever is already packed

I don't understand this - it isn't "already packed" - previously Clang packed it, but didn't warn/error the way GCC did, so you didn't know this was causing a problem (it is causing a problem - the `i` is underaligned, you leak a pointer/reference to that member that doesn't carry that underalignment information with it, so your code is likely to be incorrect whenever that pointer/reference is used (the underaligned access will be surprisingly non-atomic on x86, and crash on arm - maybe other bad things will happen elsewhere (possible that compiler optimizations could strip the low bits out of the pointer, leading to the pointer pointing to some other data)))

> I wouldn't be surprised if one day Clang would disallow also that, and the work-around of adding explicit packing will stop working even for Clang.

Agreed - though I think that all points to your code being incorrect, though?




So, I guess the summary I'd give:

This original code you have compiled correctly (in the sense that it didn't lead to underaligned integers) on GCC, with a sizeof(Bar) of 16, not what you wanted perhaps - but it is what you were getting on GCC, so I guess it's something that's acceptable to you?

The code compiled problematically on Clang previously, though produced a more compact size (10), and now it doesn't do that anymore - and it's revealed the bug (well, a combination of trying to placate Clang and then asking for the same thing from GCC caused GCC to diagnose the problem).

You can ask Clang for the old behavior with `-fclang-abi-compat=15` but I wouldn't recommend it.

**Comment by stsp** [comment]
> But with GCC you weren't getting the packing you wanted previously anyway (you were getting > sizeof(Bar) == 16, right? - so, if under GCC you were OK

No, not OK. My project was clang-specific
forever, exactly because of that. In this
particular case I am only pointing that the
_compilation_ now fails  under gcc, which
means it may soon do the same under clang
as well.

> I don't understand this - it isn't "already packed" - previously Clang packed it

But why does it need packing if in my
example it consists of a single `int`?
I think it is already "packed" and the
`-Wpacked` warning was perfectly correct!
Now you propose to change the `-Wpacked`
instead... :( But why, if the single `int` doesn't
need to be packed? What's the rationale of
forcing people to pack that?

> you leak a pointer/reference to that member that doesn't carry that underalignment
> information with it

In a sense, yes, and this may be the problem
in theory. But please, how is that different with
the POD members, where the same is still
allowed?
If you are to solve that consistently, wouldn't
you instead require to mark the resulting pointer
with some "unaligned" attribute or alike?

> crash on arm

On some old arm perhaps, as AFAIK the
armv8 allows unaligned accesses. On the
arches that do not allow this, you would
get the fixup from the kernel, but not a crash.

> Agreed - though I think that all points to your code being incorrect, though?

In a sense of gcc compatibility - perhaps.
I always thought gcc is buggy on that, and
clang was very good. What is the reason
of a change, besides the gcc compatibility,
which leads to another incompatibility, as
shown in this example?

> getting on GCC, so I guess it's something that's acceptable to you?

No. And in this ticket I only show that the
attempt to solve one gcc compatibility leads
to another one, and adds real problems to
people who used clang as their primary
compiler for decades.

> You can ask Clang for the old behavior with -fclang-abi-compat=15 but I wouldn't recommend it.

This may disappear in the future.
Can you add a more prominent switch, like
`-fnon-pod-packing` or alike?

**Comment by dwblaikie** [comment]
> > But with GCC you weren't getting the packing you wanted previously anyway (you were getting > sizeof(Bar) == 16, right? - so, if under GCC you were OK
> 
> No, not OK. My project was clang-specific forever, exactly because of that.

OK, then you can ask for the old ABI if you like, as mentioned.

> In this particular case I am only pointing that the _compilation_ now fails under gcc, which means it may soon do the same under clang as well.

The reason it fails under GCC would apply equally to the the pre-change Clang 15 ABI.

> > I don't understand this - it isn't "already packed" - previously Clang packed it
> 
> But why does it need packing if in my example it consists of a single `int`? I think it is already "packed" and the `-Wpacked` warning was perfectly correct! Now you propose to change the `-Wpacked` instead... :( But why, if the single `int` doesn't need to be packed? What's the rationale of forcing people to pack that?

I'm not really trying to rationalize it - I'm trying to match GCC's behavior because this is the ABI - if Clang doesn't implement it, then Clang and GCC are not ABI compatible - and their compatibility is an explicit goal of the Clang project.

I can speculate about why GCC chose to do this - but it's fairly inconsequential - even if it's a bad reason, we'll still implement it in Clang for compatibility.

The particular reason I came across this was that a customer of mine packed a struct that included a `std::atomic` member, silently underaligning that member and making the atomic operations no longer atomic, this lead to a significant production issue. As part of that investigation we found that if we'd correctly implemented the GCC ABI in this regard, that wouldn't've been a problem.

But that's just the proximal reason we came across it, ultimately GCC ABI compatibility except in the case of severe bugs in GCC (where they'd be willing to change the ABI themselves) is mandatory for Clang.

> > you leak a pointer/reference to that member that doesn't carry that underalignment
> > information with it
> 
> In a sense, yes, and this may be the problem in theory. But please, how is that different with the POD members, where the same is still allowed? If you are to solve that consistently, wouldn't you instead require to mark the resulting pointer with some "unaligned" attribute or alike?

Packing is really developer hostile - we (Clang, GCC, etc) try our best to diagnose cases we think are more likely to be buggy than helpful, but it's very hit-or-miss.

Looks like the error from GCC isn't about pod V non-pod: https://godbolt.org/z/MbEvffn3h it's about creating a reference to the underaligned elements - direct accesses through expressions that still "See" the packed/underalignment are OK, because the compiler can produce specific instructions/etc for those cases, but if the underalignment leaks into a reference or pointer, you're violating the language requirements that those pointers are suitably aligned and it's possible other optimizations will break the code (they could depend on the low bits of the pointer being zero, for instance).

The choice to not force packing of a pod struct member V a non-pod struct member - I think that's probably motivated because a non-pod likely has member functions and other things that act on members and assume they are aligned, etc. Essentially you're leaking a pointer (to member functions) to the struct that doesn't keep the "packed"/underaligned property - so you end up accessing the underaligned structure without knowing it.

> > crash on arm
> 
> On some old arm perhaps, as AFAIK the armv8 allows unaligned accesses. 

At least at a quick google, "except for exclusive or ordered ... allows unaligned access" - so, if you're doing atomic ops this could still crash, I guess? (which is the use case where I came across this layout problem - well, not on ARM, but if it had been, at least the user would've got a crash instead of silently non-atomic operation, I suppose)

> On the arches that do not allow this, you would get the fixup from the kernel, but not a crash.

I'm not sure what you mean by the latter ("fixup from the kernel") - my understanding is that this silently becomes non-atomic and there's no kernel involved here, I think - it's just the instruction running on the processor...  

> > Agreed - though I think that all points to your code being incorrect, though?
> 
> In a sense of gcc compatibility - perhaps. I always thought gcc is buggy on that, and clang was very good. What is the reason of a change, besides the gcc compatibility, which leads to another incompatibility, as shown in this example?

What "other incompatibility" are you referring to? The Clang warning lead to adding the attribute that gets the same layout on Clang and GCC, and then leads to GCC warning on certain uses of that layout - I don't think that's "another incompatibility" as much as it is "if you use this layout, GCC will tell you it's problematic" - if you decide you want the layout in Clang but not in GCC, then it makes sense you could use some preprocessor conditionals to ask for different layouts in each compiler. But the error GCC is providing is equally applicable to this layout, no matter how you get to it/which compiler you get it from (whether it's from Clang 15 or 16, for instance - it's still a problem).

> > getting on GCC, so I guess it's something that's acceptable to you?
> 
> No. And in this ticket I only show that the attempt to solve one gcc compatibility leads to another one, and adds real problems to people who used clang as their primary compiler for decades.

OK

> > You can ask Clang for the old behavior with -fclang-abi-compat=15 but I wouldn't recommend it.
> 
> This may disappear in the future. Can you add a more prominent switch, like `-fnon-pod-packing` or alike?

It won't disappear - if you don't care about GCC ABI compatibility, using a specific `-fclang-abi-compat` flag is what I'd recommend you use. This will protect you from future Clang ABI fixes as well. Rather than having to go through this sort of discussion every time Clang gets a bugfix for a GCC ABI incompatibility.

**Comment by stsp** [comment]
> The particular reason I came across this was that a customer of mine packed a struct that
> included a std::atomic member

Why it happened to be non-POD?
I don't think atomic is a class that inherits
some other, or is it?
Its more likely that atomic had an explicit
alignment attribute, and maybe clang
allowed packing to silently ignore the
alignment attribute?

> Essentially you're leaking a pointer (to member

To me it looks like a property of a pointer,
so maybe you wanted to mark it as unaligned
with some attribute. That could even affect
the code generation, and actually generate
the correct code for making an unaligned
access (by splitting it into multiple byte-accesses
for example). So that would at least fix things.

> I'm not sure what you mean by the latter ("fixup from the kernel") - my understanding is that
> this silently becomes non-atomic and there's no kernel involved here, I think - it's just the
> instruction running on the processor...

What you call "the crash" is actually the
exception that some RISC processors raise
on an unaligned access (or even x86 ones
do, if you enable the explicit alignment check
there). That exception is trapped by the kernel,
which then looks up the instruction that caused
that, and emulates it. This is what I call a
"kernel fixup". Such things are of course non-atomic
wrt another CPU cores, but I think your problem
with atomic was because of ignored explicit alignment.

> It won't disappear - if you don't care about GCC ABI compatibility, using a specific
> -fclang-abi-compat flag is what I'd recommend you use.

OK I take your word on that it won't disappear...
Though I'd feel much more comfortable with a
dedicated switch, because in this case it may
be regarded as a feature that the users of new
API may want to use. That's a really good feature.
I know that because I was filing the gcc bugs
about it and other people were joining the ticket
saying they want it too.

**Comment by frederick-vs-ja** [comment]
> Why it happened to be non-POD?
> I don't think atomic is a class that inherits
> some other, or is it?

`std::atomic` has no eligible copy/move constructor or copy/move assignment operator (either deleted or not declared), so it's not trivially copyable, trivial, or POD.

**Comment by stsp** [comment]
Thanks for clarifying.
In that case I suspect this fix won't fix eg
sig_atomic_t and other atomic variations.
It just happened to fix one.
Marking all atomic types with explicit
alignment would fix the problem w/o any
compiler changes.

**Comment by stsp** [comment]
https://en.cppreference.com/w/cpp/named_req/StandardLayoutType

What do you think of requiring StandardLayout for
packing, instead of the POD?
I am not sure what is "POD for the purpose of layout"
that you referred to in some other thread, but
its probably equal to StandardLayout?

**Comment by stsp** [comment]
https://github.com/itanium-cxx-abi/cxx-abi/issues/66#issuecomment-426831099
```
A type is considered a POD for the purposes of layout if (ignoring cv-qualifications) it is a scalar type, an array of whose element type is considered POD for the purposes of layout, or a class type that has:

    no user-provided or inherited constructors,
    no user-provided copy or move assignment operators,
    no user-provided destructors,
    no private or protected non-static data members,
    no non-static reference members,
    no bit-fields whose declared width is wider than the declared type of the bit-field,
    no potentially-overlapping non-static data members,
    no base classes,
    no virtual functions,

and for which all non-static data members are of types considered POD for the purposes of layout.
```

Is this a valid description of a "POD for the purpose of layout"?
If so - it doesn't seem to even require the
type to be trivial.
Yes, I've seen also many other definitions of
"POD for the purpose of layout", even those
that explicitly require PODness...
But if we use that definition of
"POD for the purpose of layout", then all problems
go away completely.

**Comment by stsp** [comment]
Also see https://github.com/itanium-cxx-abi/cxx-abi/issues/66#issuecomment-439161363
which asks:
```
And just to confirm: deleted constructors/destructors/move-assignment operators also won't
make a type non-pod-for-layout, correct? Including if they're implicitly deleted by the
declaration of a move assignment operator.
```

and https://github.com/itanium-cxx-abi/cxx-abi/issues/66#issuecomment-439371331
which replies:
```
special members that are implicitly defined as deleted are not "user-provided".
```

So if I understand these guys correctly,
"POD for the purpose of layout" doesn't
need to be a "trivial" type, hence it doesn't
need to be POD.
If this is correct, then all problems are gone!

CC to @jwakely @jyknight @zygoloid

**Comment by zygoloid** [comment]
The root cause of the problem here is that you want an `int&` to bind to a misaligned `int`, and that's not going to work reliably. For a long-term fix, one reasonable thing to do would be to tell the compiler that you're binding the reference to a misaligned `int`. For example:

```
#include <cstdio>
#define PACKED __attribute__((packed))

typedef __attribute__((aligned(1))) int packed_int;

class PACKED Foo {
    int i;
public:
    packed_int& get_ref() { return i; }
};

class PACKED Bar {
public:
    unsigned char c;
    Foo p;
    unsigned char c1;
    int c2;
    int get_ref() { return p.get_ref(); }
};

static_assert(sizeof(Bar) == 10);

int main()
{
  Bar b;
  printf("sizeof(b)=%zi\n", sizeof(b));
  return b.get_ref();
}
```

Unfortunately, GCC rejects this at the moment, but [I think that's just a GCC implementation bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=109654).

**Comment by stsp** [comment]
Richard, yes, this is exactly what I proposed in
https://github.com/llvm/llvm-project/issues/62358#issuecomment-1524432437

But we are discussing multiple problems here
in a single ticket. The outstanding problem currently
is that clang-16 agrees to pack PODs and refuses
to pack non-PODs. It is very reasonable to refuse
packing non-std-layout types, so I looked as what
"POD for the purpose of layout" actually is.
And according to your discussion I referenced
above, "POD for the purpose of layout" doesn't
require the type to be "trivial" or POD. Which
basically confirms my theory that only
non-std-layout types should be refused from
packing.

Now could you please clarify if
"POD for the purpose of layout" requires the
type to be trivial? Because if not - the problem
doesn't even exist.

**Comment by stsp** [comment]
```
#include <stdio.h>
#include <type_traits>

struct dhdr {
  int a;
};

template<typename T>
class SymWrp : public T {
public:
    SymWrp() = default;
};

struct lol {
  unsigned char _l;
  SymWrp<struct dhdr> _n;
  unsigned char _nj;
} __attribute__((packed));

int main()
{
    static_assert(std::is_standard_layout<SymWrp<struct dhdr>>::value);
    static_assert(std::is_trivial<SymWrp<struct dhdr>>::value);
    static_assert(std::is_pod<SymWrp<struct dhdr>>::value);
    printf("%zi\n", sizeof(struct lol));
    return 0;
}
```

Now I inserted the static asserts to the example,
and none triggers. And yet clang-16 says this:
```
$ clang++-16 -Wpacked -Wpacked-non-pod -o p_clng pod.cpp
pod.cpp:16:23: warning: not packing field '_n' as it is non-POD for the purposes of layout [-Wpacked-non-pod]
  SymWrp<struct dhdr> _n;
```

So now I am completely confused... :(
If the type is std and trivial and POD, then
how come its still a "non-POD for the purposes of layout"?

**Comment by zygoloid** [comment]
> If the type is std and trivial and POD, then
> how come its still a "non-POD for the purposes of layout"?

Because "POD for the purposes of layout" has nothing to do with the C++11 definition of "POD" (standard layout and trivially-copyable). It's based on the C++03 definition of "POD", with amendments to accommodate features that didn't exist in C++03 (move operations, defaulted and deleted functions, inherited constructors, `[[no_unique_address]]`).

`SymWrp<dhdr>` is not "POD for the purposes of layout" because it has a base class.

**Comment by stsp** [comment]
Richard, yes, I already realized that fact.
That's because, unlike normal POD,
this POD (and C++03 POD) requires
the class to be aggregate.
I however do propose to evaluate the
possibility (maybe an optional one) of
using a C++11 POD definition for that.
If gcc compatibility is The Thing (tm),
then some -fpacking-pod-cpp11 or alike
can probably be added.

**Comment by zygoloid** [comment]
I don't think a `-f` flag to introduce an ABI variant is a good path forward here. We regrettably support `-fclang-abi-compat=...` because targeting "the same ABI that Clang used to produce" is an important use case, but we generally don't want an explosion of different ABI variations. (For example, this could lead to your code failing at runtime if you link against a library built with a different setting for the `-f` flag -- and that might even be the C standard library.)

We could add a new language feature for this -- such as a new attribute whose semantics could be defined to suit your use case -- but the bar for doing so is very high. Instead, I'd suggest you look for a way to get the effect you're looking for with features that already exist. For example, you could use explicit alignment control instead of the `packed` attribute:

```
typedef __attribute__((aligned(1))) int packed_int;

class Foo {
    packed_int i;
public:
    packed_int& get_ref() { return i; }
};
```

**Comment by stsp** [comment]
Yes, -fclang-abi-compat can lead to a
problems. Which is why I dont want to
use it. But the specific -fpacking-xxx
will not lead to any problems. Thats
because the clang16's default is to warn
and ignore the user-supplied packing
attribute! Ignore something written by
the programner, is a very strange decision
and from my pov its an attempt to
transform an ill-formed by compiler's
opinion program into an ill-formed by
programmer's opinion program. And
it produces the warning anyway. So
such situation is definitely not possible
with glibc or any other popular lib - they
won't provide the code where compiler
needs to ignore something with a warning.
So in this particular case I think the -f
option to not ignore the user-specified
attribute is safe and sensible.

The explicit alignment control you suggest,
is not helping at all. It doesn't work on gcc,
and on clang-16 its not needed as clang-16
still allows the unaligned references.
Also note that these are orthogonal
problems: if you add -fpacking-xxx opt
and clang will eventually disallow the
unaligned references, then I will need
to use both: an option and an aligned
attr for a reference, but we are not there
yet.

**Comment by jyknight** [comment]
Please do note that Clang _already_ disallows unaligned references. It simply fails to warn you about the undefined-behavior.

**Comment by stsp** [comment]
Thanks for a note.
I think this comment explains:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=36566#c14
It seems to be saying gcc errors
out even when the object is actually
properly aligned, but resides in a
packed struct so _could_ be mis-aligned
but wasn't.
Do you think clang will port in this bug
for gcc compatibility, or will it warn only
when an object is _actually_ mis-aligned?

**Comment by dwblaikie** [comment]
> Yes, -fclang-abi-compat can lead to a problems. Which is why I dont want to use it. 

What problems are you concerned `-fclang-abi-compat` might lead to? This flag has long-term/indefinite support & will protect you from /other/ ABI fixes that might occur and disrupt your layout preferences in the future too.

> Undoubtedly, I think you can even error on that the same way gcc does. But how about not to warn when alignment is not violated even though packing is used? Its undoubtedly a bug, so it won't be ported to clang, right?

These issue discussions are getting a bit long-winded, it's probably best not to include too much speculation about bugs on top of bugs. But, yes, if/when at some arbitrary (not soon - no one's stepping up to implement this) point in the future if/when someone implements the unaligned access warning in clang, we'd endeavor to avoid bugs like this.

**Comment by stsp** [comment]
> What problems are you concerned -fclang-abi-compat might lead to?

As Richard points out, it may eventually
became incompatible with anything, eg
libstdc++. Who knows how ABI evolves
in the future, how well the new C++
standards are going to be supported on
clang-15 ABI and so on.

OTOH the explicit option for that problem
is not going to be incompatible with anything,
as it limits not the entire ABI to what it was
in clang-15, but rather allows the explicitly
written packing attribute to not get ignored.
The only way to get 2 incompatible objects,
is to ignore the gcc warning, which should
be an error actually, because ignoring something
written by a programmer, is an unusual way
of "fixing" the program. So it is simply not
possible to have some lib like libstdc++ with
such a problem. Hence the real incompatibilities
are not possible if such an option is added.
Which is why it looks like the way forward
to me, and clang-15 ABI option looks like a
temporary solution at best.

**Comment by jyknight** [comment]
I'm currently inclined to close this working-as-intended.

Yes, the ABI changed -- to be compatible with GCC. That was an intentional change. If your concern was compatibility between old and new clang-compiled code, then you can use `-fclang-abi-compat`. But now it sounds like you're asking for something beyond "compatibility" -- that you simply _prefer_ the old ABI, and want to keep that behavior forever. That needs justification.

As discussed, the example at the top is UB (in both GCC and Clang). Is there a real use-case for such flag which doesn't involve UB? (And not just a made-up example showing that the size of an object can change --  why is this _important_?)

**Comment by zygoloid** [comment]
> In this example `p` is in the beginning of the `Bar`, so its still properly aligned, even if packed. Is this an UB too?

There's no guarantee that the `int` here is properly aligned. The alignment of `Bar b` is 1, so `b.p.n` might not be properly aligned for an `int`, so binding an `int&` to it forms a misaligned reference. So yes, this example has undefined behavior (if `b` happens to not be 4-byte aligned, which is permitted).

**Comment by stsp** [comment]
That's a good point, thanks.
I do not suppose removing packed from
`Foo` changes anything in what you said
above, right?

**Comment by stsp** [comment]
Would you like to answer
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=36566#c14
with that your point?
Maybe they will find the "counter-attack".

**Comment by stsp** [comment]
So I guess in my code I need to mark the
references with `aligned(1)` as you suggested
in another thread. But currently gcc ignores
that, and clang doesn't require that, so perhaps
for now I can't avoid this UB. But I agree this
would be needed if clang starts to enforce an
alignment of references.

**Comment by dwblaikie** [comment]
> I do not suppose removing packed from `Foo` changes anything in what you said above, right?

If you remove packed from `Foo`, then `Foo`s alignment and size requirements are respected (& GCC warns you that, for this field, the packed attribute will not apply/will be ignored - though it still packs the rest of the struct, but can't lower the alignment requirement of the struct below the alignment requirement of `Foo`) - so there's no unaligned reference binding anymore: https://godbolt.org/z/EcjrsavE1

**Comment by stsp** [comment]
OK in this case I agree that the problem
boils down only to adding a few extra
packings and ignoring the broken -Wpacked
diagnostic. I wish to avoid having to do that,
but the problem mentioned in this ticket is
not valid, and unless clang is not going to
copy the gcc behavior but would allow marking
the references with alignment of 1, or will
provide some other means to not get into a
hard error as in gcc, then this ticket can be
closed.
